
-- Example texture file

--!strict
--!optimize 2
--!native

if _G.ROBLOX then
	local Asset = game:GetService 'AssetService'
	local TextureContent = Content.fromAssetId(126128421498657)

	local Image = Asset:CreateEditableImageAsync(TextureContent)
	local ImageSize = Image.Size
	local ImageData = Image:ReadPixelsBuffer(Vector2.zero, ImageSize)

	return {
		width = ImageSize.x,
		height = ImageSize.y,
		data = ImageData
	}
else
	return {
		width = 8,
		height = 8,
		data = buffer.fromstring(
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF` ..
			`\xf5\x87\x42\xFF\x42\x87\xf5\xFF\x42\x87\xf5\xFF\xf5\x87\x42\xFF`
		)
	}
end